packages/eslint-plugin/src/rules/prefer-readonly.ts

```diff
@@ -229,13 +229,48 @@ export default createRule<Options, MessageIds>({
             }
           })();
 
+          const typeAnnotation = (() => {
+            if (violatingNode.type || !violatingNode.initializer) {
+              return null;
+            }
+
+            if (esNode.type !== AST_NODE_TYPES.PropertyDefinition) {
+              return null;
+            }
+
+            const initializerType = checker.getTypeAtLocation(
+              violatingNode.initializer,
+            );
+
+            const violatingType = checker.getTypeAtLocation(violatingNode);
+
+            // if the RHS is a literal, its type would be narrowed, while the
+            // type of the initializer (which isn't `readonly`) would be the
+            // widened type
+            if (initializerType === violatingType) {
+              return null;
+            }
+
+            if (!tsutils.isLiteralType(initializerType)) {
```

Comment:
[Bug] It's not just literal types though, it could also be [unions of literals (playground)](https://deploy-preview-10552--typescript-eslint.netlify.app/play/#ts=5.7.2&fileType=.ts&code=MYGwhgzhAEAqCmEAu0DeAoa0AOAnAlgG5hLw64D220AvNAORj0Dc6m0wFAdsrgK7AkFXAAoAlGnZYkAC3wQAdHiq0GAIxZTos%2BUsrU69YJqwBfdOaA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIXRRe0EdEjgwAXxBKgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false):  

```ts
class Test {
  private prop = 'a';

  constructor() {
    this.prop = 'b';
    this.prop = 'c';
  }
}
```

...or a [template literal (playground)](https://deploy-preview-10552--typescript-eslint.netlify.app/play/#ts=5.7.2&fileType=.ts&code=C4TwDgpgBAMhzAgJwAoEMkQHYGcoF4oADACgBIBvAcjSqgB8oqAjKgXwEoiAobgEwgBjADYZoggPa5gUMGNwAuWPESp5OANy8RaHHgAqEHDIrcospAEsAbmkQWJYAkxJoOVLWaiTpSAK6CwBJIJBxQpubmwAAWljgAdGBIjs5ymLha5mzc2UA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIXRRe0EdEjgwAXxBKgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false):

```ts
type LetterParens = `(${'a' | 'b'})`

declare const parens: LetterParens;

class Test {
  private prop = '(a)';

  constructor() {
    this.prop = parens;
  }
}
```

I think this'll have to use the assignability API.

Reply:
I'm confused. Don't these examples currently have the desired behavior?

Reply:
I suppose from a certain point of view they do. Hmm. I was thinking that the right fixed type here would be a union like `'a' | 'b' | 'c'` or `LetterParens`. But I suppose they do work fine with the more broad `| string`. Especially if it means we don't have to use the more nuanced assignability API.

Ok 👍 never mind from me then!

Reply:
Yeah, my thinking was that it positively should be `string` and not `'a' | 'b' | 'c'` since that's the type that the member had before the fix (due to literal widening). So putting the widened type in the fix, and leaving it up to the user to explicitly narrow it down further if they want seems certainly least surprising to me. That being said, we could still turn it into a suggestion instead of a fix if we wanted to draw extra attention to it, since it might be something the user might wish to further refine? 🤷‍♂️

Reply:
Heh I don't think I've ever seen a rule that provides a fix and a suggestion in the same place. Kind of a weird thing to do. I think it's fine as is.